File: LeetCode/javascript/lc_0145.js
/**
 * 递归
 * @param {TreeNode} root
 * @return {number[]}
 */
var postorderTraversal = function (root) {
    let result = [];

    function traversal(node) {
        if (node) {
            if (node.left) traversal(node.left);
            if (node.right) traversal(node.right);
            result.push(node.val);
        }
    }

    traversal(root);
    return result;
};


/**
 * 迭代 - 使用visited记录是否访问过右孩子
 * @param {TreeNode} root
 * @return {number[]}
 */
var postorderTraversal = function (root) {
    let result = [];

    let stack = [];
    let curNode = root;
    let visited = null;
    while (stack.length > 0 || curNode) {
        while (curNode) {
            stack.push(curNode);
            curNode = curNode.left;
        }

        curNode = stack.pop();
        if (curNode.right == null || curNode.right == visited) {
            // 无右孩子 | 右孩子已访问 -> 开始回退
            result.push(curNode.val);
            visited = curNode;
            curNode = null;
        } else {
            // 有右孩子，回栈，进入右孩子
            stack.push(curNode);
            curNode = curNode.right;
        }
    }

    return result;
};


/**
 * Morris
 * @param {TreeNode} root
 * @return {number[]}
 */
var postorderTraversal = function (root) {
    let result = [];

    let curNode = root;
    while (curNode) {
        if (curNode.left) {
            let mostRight = curNode.left;
            while (mostRight.right && mostRight.right != curNode) {
                mostRight = mostRight.right;
            }

            if (mostRight.right == null) {
                mostRight.right = curNode; // mark
                curNode = curNode.left;
                continue;
            } else {
                // 左子树遍历完毕
                mostRight.right = null; // resume
                _collectResult(curNode.left);
            }
        }
        curNode = curNode.right;
    }
    _collectResult(root);

    function _collectResult(head) {
        let tail = _reverseLinkedList(head); // reverse

        while (tail) {
            result.push(tail.val);
            tail = tail.right;
        }

        _reverseLinkedList(tail); // resume
    }

    function _reverseLinkedList(node) {
        let pre = null;
        while (node) {
            let nextNode = node.right;
            node.right = pre;
            pre = node;
            node = nextNode;
        }

        return pre; // new head
    }

    return result;
};

File: LeetCode/cpp/lc_0145.cpp
/**
 * @file     lc_0145.cpp
 * @brief    https://leetcode-cn.com/problems/binary-tree-postorder-traversal/
 * @author   YongDu
 * @date     2021-09-09
 */
class Solution {
  public:
    vector<int> result; // 结果集

    vector<int> postorderTraversal(TreeNode* root) {
        if (nullptr == root)
            return result;

#if 0
        // 1. 递归版
        if(root) {
            postorderTraversal(root->left);
            postorderTraversal(root->right);
            result.emplace_back(root->val);
        }
        return result;
#endif

        // 2. 迭代版
        stack<TreeNode*> stk;
        TreeNode* node = root;
        stk.push(node);
        while (!stk.empty()) {
            node = stk.top();
            stk.pop();
            result.emplace_back(node->val);
            if (node->left)
                stk.push(node->left);
            if (node->right)
                stk.push(node->right);
        }

        reverse(result.begin(), result.end());
        return result;
    }
};
